### PR TITLE
feat:Advantages of TypeScript(#39520)

### PR DESCRIPTION
**Pull request by Rajshrishrimalshah**
please review my assignment.

### CODE REVIEW

File: extra/constant.js [hunk 1]
```diff
@@ -9,3 +9,25 @@ export const ROLL_OF_HEAD_TRAINER = "head-trainer";
 
 
 
+export const users = [
```
Comment (SunilSuccessive) [inline extra/constant.js:10]: Please specify the variable types everywhere because you are writing ts not js

File: extra/index.js [hunk 2]
```diff
@@ -1,6 +1,22 @@
+ import { diamond,equilateral } from "./patterns"
```
Comment (SunilSuccessive) [inline extra/index.js]: please use proper spacing

File: extra/index.js [hunk 3]
```diff
@@ -1,6 +1,22 @@
...
 
-import hasPermission from './extra/utils/permissions'
+ import {PERMISSION_ALL,PERMISSION_READ,PERMISSION_WRITE,PERMISSION_DELETE,
```
Comment (SunilSuccessive) [inline extra/index.js]: Please use proper spacing

File: extra/patterns/diamond.js [hunk 4]
```diff
@@ -1,56 +1,31 @@
...
-
-			process.stdout.write("\n"); // after each row completion prints 'NEWLINE
-
-		}	}
-
-	}
-
-let max = process.argv[2];
-star(max); // Function Calling
+export const diamond = (max) => {
```
Comment (SunilSuccessive) [inline extra/patterns/diamond.js]: please define types of function everywhere

**Review by SunilSuccessive — CHANGES_REQUESTED** [review]

File: extra/index.js (hunk 3 above)
Comment (Rajshrishrimalshah) [reply extra/index.js]: done

File: extra/index.js (hunk 2 above)
Comment (Rajshrishrimalshah) [reply extra/index.js]: done

File: extra/constant.js (hunk 1 above)
Comment (Rajshrishrimalshah) [reply extra/constant.js:10]: done.

File: extra/patterns/diamond.js (hunk 4 above)
Comment (Rajshrishrimalshah) [reply extra/patterns/diamond.js]: done.

File: extraTs/constant.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+export const PERMISSION_WRITE: string = "write";
+export const PERMISSION_DELETE: string = "delete";
+export const MODULE_USER: string = "getUsers";
+export const ROLE_OF_TRAINER: string = "trainer";
+export const ROLE_OF_TRAINEE: string = "trainee";
+export const ROLL_OF_HEAD_TRAINER: string = "head-trainer";
+
+
+
+export const users = [
```
Comment (SunilSuccessive) [inline extraTs/constant.ts]: Define types of users object

File: extraTs/interfaces.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+export interface Ipattern {
```
Comment (SunilSuccessive) [inline extraTs/interfaces.ts]: P should be capital in Ipattern.
Naming Convention for interface: -
 - Use pascal case
 - Start with I
 - Use meaningful variable or method name.
 - e.g., method name is pattern and interface should be IPattern.

Comment (Rajshrishrimalshah) [reply extraTs/interfaces.ts]: done.

File: extraTs/constant.ts (hunk 5 above)
Comment (Rajshrishrimalshah) [reply extraTs/constant.ts]: done.

**Pull request merged**